File: src/main/java/no/nav/pto_schema/kafka/json/JsonSchemaLocator.java
package no.nav.pto_schema.kafka.json;

import lombok.SneakyThrows;
import org.everit.json.schema.Schema;
import org.everit.json.schema.loader.SchemaLoader;
import org.json.JSONObject;
import org.json.JSONTokener;

import java.io.InputStream;
import java.nio.charset.StandardCharsets;

public class JsonSchemaLocator {

    public static Schema getKafkaSchema(Class<?> schemaTypeClass) {
        return toSchema(getKafkaSchemaStr(schemaTypeClass));
    }

    public static String getKafkaSchemaStr(Class<?> schemaTypeClass) {
        String kafkaSchemaResourcePath = createKafkaSchemaResourcePath(schemaTypeClass);
        return getResourceFileAsString(kafkaSchemaResourcePath);
    }

    public static String createKafkaSchemaResourcePath(Class<?> kafkaSchemaClass) {
        String schemaName = getSchemaName(kafkaSchemaClass);
        return "kafka/json/topic/" + schemaName;
    }

    /**
     * Creates a schema name from a class
     * @param schemaClass class which will be used to create schema name
     * @return schema name of class. example: MyTopicV1.class -> my-topic-v1-schema.json
     */
    public static String getSchemaName(Class<?> schemaClass) {
        String className = schemaClass.getSimpleName();
        StringBuilder schemaNameBuilder = new StringBuilder();

        char[] chars = className.toCharArray();

        for (int i = 0; i < chars.length; i++) {
            char ch = chars[i];

            if (Character.isUpperCase(ch) && i > 0) {
                schemaNameBuilder.append('-');
            }

            schemaNameBuilder.append(Character.toLowerCase(ch));
        }

        return schemaNameBuilder.toString() + "-schema.json";
    }

    private static Schema toSchema(String schemaJson) {
        return SchemaLoader.load(new JSONObject(new JSONTokener(schemaJson)));
    }

    @SneakyThrows
    private static String getResourceFileAsString(String resourcePath) {
        ClassLoader classLoader = JsonSchemaLocator.class.getClassLoader();

        try (InputStream resourceStream = classLoader.getResourceAsStream(resourcePath)) {
            if (resourceStream == null) {
                throw new IllegalArgumentException("No resource found for " + resourcePath);
            }

            return new String(resourceStream.readAllBytes(), StandardCharsets.UTF_8);
        }
    }


}

File: src/main/kotlin/no/nav/pto_schema/kafka/json/JsonSchemaLocator.kt
package no.nav.pto_schema.kafka.json

import org.everit.json.schema.Schema
import org.everit.json.schema.loader.SchemaLoader
import org.json.JSONObject
import org.json.JSONTokener
import java.nio.charset.StandardCharsets

object JsonSchemaLocator {

    @JvmStatic
    fun getKafkaSchema(schemaTypeClass: Class<*>): Schema =
        toSchema(getKafkaSchemaStr(schemaTypeClass))

    @JvmStatic
    fun getKafkaSchemaStr(schemaTypeClass: Class<*>): String {
        val kafkaSchemaResourcePath = createKafkaSchemaResourcePath(schemaTypeClass)
        return getResourceFileAsString(kafkaSchemaResourcePath)
    }

    @JvmStatic
    fun createKafkaSchemaResourcePath(kafkaSchemaClass: Class<*>): String {
        val schemaName = getSchemaName(kafkaSchemaClass)
        return "kafka/json/topic/$schemaName"
    }

    /**
     * Creates a schema name from a class
     * @param schemaClass class which will be used to create schema name
     * @return schema name of class. example: MyTopicV1.class -> my-topic-v1-schema.json
     */
    @JvmStatic
    fun getSchemaName(schemaClass: Class<*>): String {
        val className = schemaClass.simpleName
        val schemaNameBuilder = StringBuilder()

        className.forEachIndexed { i, ch ->
            if (ch.isUpperCase() && i > 0) {
                schemaNameBuilder.append('-')
            }
            schemaNameBuilder.append(ch.lowercaseChar())
        }

        return "$schemaNameBuilder-schema.json"
    }

    private fun toSchema(schemaJson: String): Schema =
        SchemaLoader.load(JSONObject(JSONTokener(schemaJson)))

    private fun getResourceFileAsString(resourcePath: String): String {
        val classLoader = JsonSchemaLocator::class.java.classLoader

        classLoader.getResourceAsStream(resourcePath).use { resourceStream ->
            requireNotNull(resourceStream) { "No resource found for $resourcePath" }
            return String(resourceStream.readAllBytes(), StandardCharsets.UTF_8)
        }
    }
}

